weave/trace_server/execution_runner/cross_process_trace_server.py

```diff
@@ -0,0 +1,540 @@
+    _secret_fetcher_context,
+]
+
+# Default timeout for request/response cycles
+TIMEOUT_SECONDS = 60.0
+
+
+class RequestWrapper(TypedDict):
+    """Wrapper for requests sent to the worker thread."""
+
+    method: str
+    request: Any
+    request_id: str
+
+
+class ResponseWrapper(TypedDict):
+    """Wrapper for responses from the worker thread."""
+
+    request_id: str
+    result: Optional[Any]
+    error: Optional[str]
+
+
+class CrossProcessTraceServerArgs(TypedDict):
+    """Arguments for constructing a CrossProcessTraceServer."""
+
+    request_queue: multiprocessing.Queue
+    response_queue: multiprocessing.Queue
+
+
+class SendRequestException(Exception):
+    """Raised when an error occurs while processing a request in the worker thread."""
+
+    pass
+
+
+class CrossProcessTraceServer(TraceServerInterface):
+    """
+    A trace server implementation that delegates operations via queues.
+
+    This class acts as a proxy that sends requests through a request queue
+    and receives responses through a response queue. It doesn't know about
+    the actual trace server implementation - it just handles the
+    communication protocol.
+
+    The actual trace server runs in a separate thread (in the parent process)
+    and this class can be used from child processes to communicate with it.
+    """
+
+    def __init__(
+        self,
+        request_queue: multiprocessing.Queue,
+        response_queue: multiprocessing.Queue,
+    ):
+        """
+        Initialize the cross-process trace server.
+
+        Args:
+            request_queue: Queue for sending requests to the worker thread
+            response_queue: Queue for receiving responses from the worker thread
+        """
+        self._request_queue = request_queue
+        self._response_queue = response_queue
+
+        # Track ongoing requests for matching responses
+        self._pending_requests: dict[str, threading.Event] = {}
+        self._responses: dict[str, ResponseWrapper] = {}
+        self._request_counter = 0
+        self._lock = threading.Lock()
+
+        # Flag to track if we're shutting down - MUST be set before starting thread
+        self._shutdown = False
+
+        # Start response handler thread
+        self._response_handler_thread = threading.Thread(
+            target=self._handle_responses, daemon=True
+        )
+        self._response_handler_thread.start()
+
+    def _generate_request_id(self) -> str:
+        """Generate a unique request ID."""
+        with self._lock:
+            self._request_counter += 1
+            return f"req_{self._request_counter}"
+
+    def _handle_responses(self) -> None:
+        """
+        Handle responses from the worker thread.
+
+        This method runs in a separate thread and continuously polls the
+        response queue for incoming responses, matching them with pending
+        requests.
+        """
+        while not self._shutdown:
```

Comment:
do we need a polling interval here? how frequent is this?

Reply:
the timeout is what controls the frequency